### ISSUE TITLE
Refactorer panics with 'rewriting did not complete'

### ISSUE THREAD

**Issue by TheDan64** [immunant/c2rust#137]
Running the `ownership_annotate` transform on:
```rust
#![feature(rustc_private)]

extern crate libc;

extern "C" {
    #[no_mangle]
    fn memmove(_: *mut libc::c_void, _: *const libc::c_void, _: libc::c_ulong)
     -> *mut libc::c_void;
}
pub type size_t = libc::c_ulong;
pub type __uint16_t = libc::c_ushort;
/* Key/data structure -- a Data-Base Thang. */
#[derive ( Copy , Clone )]
#[repr(C)]
pub struct DBT {
    pub data: *mut libc::c_void,
    pub size: size_t,
}
/*
 * We have to know that the key will fit, but the last entry on the page is
 * an overflow pair, so we need to shift things.
 */
unsafe extern "C" fn squeeze_key(mut sp: *mut __uint16_t, mut key: *const DBT,
                                 mut val: *const DBT) {
    let mut p: *mut libc::c_char = 0 as *mut libc::c_char;
    let mut free_space: __uint16_t = 0;
    let mut n: __uint16_t = 0;
    let mut off: __uint16_t = 0;
    let mut pageno: __uint16_t = 0;
    p = sp as *mut libc::c_char;
    n = *sp.offset(0isize);
    free_space =
        *sp.offset((*sp.offset(0isize) as libc::c_int + 1i32) as isize);
    off = *sp.offset((*sp.offset(0isize) as libc::c_int + 2i32) as isize);
    pageno = *sp.offset((n as libc::c_int - 1i32) as isize);
    off =
        (off as libc::c_ulong).wrapping_sub((*key).size) as __uint16_t as
            __uint16_t;
    *sp.offset((n as libc::c_int - 1i32) as isize) = off;
    memmove(p.offset(off as libc::c_int as isize) as *mut libc::c_void,
            (*key).data, (*key).size);
    off =
        (off as libc::c_ulong).wrapping_sub((*val).size) as __uint16_t as
            __uint16_t;
    *sp.offset(n as isize) = off;
    memmove(p.offset(off as libc::c_int as isize) as *mut libc::c_void,
            (*val).data, (*val).size);
    *sp.offset(0isize) = (n as libc::c_int + 2i32) as __uint16_t;
    *sp.offset((n as libc::c_int + 1i32) as isize) = pageno;
    *sp.offset((n as libc::c_int + 2i32) as isize) = 0i32 as __uint16_t;
    *sp.offset((*sp.offset(0isize) as libc::c_int + 1i32) as isize) =
        (free_space as
             libc::c_ulong).wrapping_sub((2i32 as
                                              libc::c_ulong).wrapping_mul(::std::mem::size_of::<__uint16_t>()
                                                                              as
                                                                              libc::c_ulong).wrapping_add((*key).size).wrapping_add((*val).size))
            as __uint16_t;
    *sp.offset((*sp.offset(0isize) as libc::c_int + 2i32) as isize) = off;
}

fn main() {
}
```

Ex via, `c2rust-refactor select ann 'crate; desc(fn | field);' \; ownership_annotate ann -- file.rs --edition 2018`

Will cause the refactorer to panic with `rewriting did not complete`:

```
thread 'main' panicked at 'rewriting did not complete', c2rust-refactor/src/rewrite/mod.rs:323:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:212
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:475
   7: std::panicking::begin_panic
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/panicking.rs:409
   8: c2rust_refactor::rewrite::rewrite
             at c2rust-refactor/src/rewrite/mod.rs:323
   9: c2rust_refactor::command::RefactorState::save_crate
             at c2rust-refactor/src/command.rs:223
  10: c2rust_refactor::main_impl::{{closure}}
             at c2rust-refactor/src/lib.rs:469
  11: c2rust_refactor::driver::run_refactoring::{{closure}}::{{closure}}::{{closure}}
             at c2rust-refactor/src/driver.rs:311
  12: rustc::ty::context::tls::with_thread_locals::{{closure}}::{{closure}}
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/librustc/ty/context.rs:1917
  13: std::thread::local::LocalKey<T>::try_with
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/thread/local.rs:257
  14: std::thread::local::LocalKey<T>::with
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/thread/local.rs:234
  15: rustc::ty::context::tls::with_thread_locals::{{closure}}
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/librustc/ty/context.rs:1909
  16: std::thread::local::LocalKey<T>::try_with
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/thread/local.rs:257
  17: std::thread::local::LocalKey<T>::with
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/thread/local.rs:234
  18: rustc::ty::context::tls::with_thread_locals
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/librustc/ty/context.rs:1901
  19: c2rust_refactor::driver::run_refactoring::{{closure}}::{{closure}}
             at c2rust-refactor/src/driver.rs:309
  20: scoped_tls::ScopedKey<T>::set
             at /cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
  21: c2rust_refactor::driver::run_refactoring::{{closure}}
             at c2rust-refactor/src/driver.rs:308
  22: scoped_tls::ScopedKey<T>::set
             at /cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
  23: syntax::with_globals::{{closure}}
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libsyntax/lib.rs:108
  24: scoped_tls::ScopedKey<T>::set
             at /cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
  25: syntax::with_globals
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libsyntax/lib.rs:107
  26: c2rust_refactor::driver::run_refactoring
             at c2rust-refactor/src/driver.rs:307
  27: c2rust_refactor::main_impl
             at c2rust-refactor/src/lib.rs:454
  28: c2rust_refactor::lib_main::{{closure}}
             at c2rust-refactor/src/lib.rs:365
  29: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/panic.rs:309
  30: std::panicking::try::do_call
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/panicking.rs:294
  31: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  32: std::panicking::try
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/panicking.rs:273
  33: std::panic::catch_unwind
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/panic.rs:388
  34: rustc_driver::report_ices_to_stderr_if_any
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/librustc_driver/lib.rs:1116
  35: c2rust_refactor::lib_main
             at c2rust-refactor/src/lib.rs:365
  36: c2rust_refactor::main
             at c2rust/src/bin/c2rust-refactor.rs:26
  37: std::rt::lang_start::{{closure}}
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/rt.rs:64
  38: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  39: std::panicking::try::do_call
             at src/libstd/panicking.rs:294
  40: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  41: std::panicking::try
             at src/libstd/panicking.rs:273
  42: std::panic::catch_unwind
             at src/libstd/panic.rs:388
  43: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  44: std::rt::lang_start
             at /rustc/929b48ec98aaff2239257574b5897f419cec2647/src/libstd/rt.rs:64
  45: main
  46: __libc_start_main
  47: _start
```

**Comment by TheDan64** [comment]
You can remove the entirety of `squeeze_key`'s body and it still happens

**Comment by rinon** [comment]
Woah, that's special. Must be something with my new comment handling. I'll take this one.

**Comment by rinon** [comment]
First, just a quick note, the select command has a double pipe for or (`||`), but it looks like that was just a typo in writing up the issue.

**Comment by rinon** [comment]
Fixed in baed0189
